#' Definition of COTAN class
#' @slot raw raw UMI count matrix ğ‘›Ã—ğ‘š (gene number Ã— cell number)
#' @slot rawNorm raw UMI count matrix divided by UDE, ğ‘›Ã—ğ‘š
#' @slot coex correlation of COTAN between genes, ğ‘›Ã—ğ‘›
#' @slot nu vector that stores the estimated UDE, size ğ‘š
#' @slot lambda vector to store the average for the gene expression, size ğ‘›
#' @slot dispetrion vector to store all
#' the negative binomial dispersion factors, size ğ‘›.
#' @slot hKGenes house-keeping genes. It is a vector to store the name 
#' of the genes with positive UMI count in every single cell of the sample
#' @slot nCells number of the cells in the sample (ğ‘š)
#' @slot metaDataset data.frame
#' @slot metaCells data.frame
#' @slot clustersCoex coex
setClass("COTAN",
         slots = c(
           raw          = "dgCMatrix",
           rawNorm      = "dgCMatrix",
           coex         = "ANY",
           cellsCoex    = "ANY",
           nu           = "vector",
           lambda       = "vector",
           dispersion   = "vector",
           hkGenes      = "vector",
           nCells       = "numeric",
           metaDataset  = "data.frame",
           metaCells    = "data.frame",
           clustersCoex = "list"
         )
)

# constructor of the COTAN CLASS
#' @export 
COTAN <- function(raw = "ANY") {
  raw <- methods::as(as.matrix(raw), "sparseMatrix")
  new("COTAN", raw = raw)
}
